### PR TITLE
Coverity fixes

### PR DESCRIPTION
**Pull request by cmouse**
### Short description
Some coverity fixes.

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [ ] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [x] added or modified unit test(s)

### CODE REVIEW

File: pdns/dbdnsseckeeper.cc [hunk 1]
```diff
@@ -144,8 +144,8 @@ bool DNSSECKeeper::clearMetaCache(const DNSName& name)
 
 void DNSSECKeeper::clearCaches(const DNSName& name)
 {
-  clearKeyCache(name);
-  clearMetaCache(name);
+  if (!clearKeyCache(name) or !clearMetaCache(name))
```
Comment (rgacogne) [inline pdns/dbdnsseckeeper.cc:147]:
```suggestion
  if (!clearKeyCache(name) || !clearMetaCache(name))
```

File: pdns/signingpipe.cc [hunk 2]
```diff
@@ -32,7 +32,9 @@ int readn(int fd, void* buffer, unsigned int len)
       if(errno == EAGAIN || errno == EINTR) {
         if(pos==0)
           return -1;
-        waitForData(fd, -1); 
+        // error handled later
+        (void)waitForData(fd, -1);
+          return -1;
```
Comment (rgacogne) [inline pdns/signingpipe.cc:37]: Why the additional `return -1` here?

**Review by rgacogne — CHANGES_REQUESTED** [review]
Thanks for this PR! I'm afraid I'm really uneasy with asserting on the return value of `chopOff()`, I'm not 100% sure a return value of `false` is an error in all these occurrences. And even if it is, it would turn a silent error that we might be handling later into a process termination, which sounds much worse.

File: pdns/distributor.hh [hunk 3]
```diff
@@ -181,7 +183,7 @@ template<class Answer, class Question, class Backend>MultiThreadDistributor<Answ
 // start of a new thread
 template<class Answer, class Question, class Backend>void MultiThreadDistributor<Answer,Question,Backend>::distribute(int ournum)
 {
-  setThreadName("pdns/distributo");
+  setThreadName("pdns/distributor");
```
Comment (rgacogne) [inline pdns/distributor.hh:186]: Pieter would know but I'm pretty sure we truncated this because there wasn't any space left.

Comment (pieterlexis) [reply pdns/distributor.hh:186]: correct!

**Comment by cmouse** [comment]
Fixed nits, dropped the assert patch.

File: pdns/dbdnsseckeeper.cc [hunk 4]
```diff
@@ -144,8 +144,8 @@ bool DNSSECKeeper::clearMetaCache(const DNSName& name)
 
 void DNSSECKeeper::clearCaches(const DNSName& name)
 {
-  clearKeyCache(name);
-  clearMetaCache(name);
+  if (!clearKeyCache(name) || !clearMetaCache(name))
+     g_log<<Logger::Warning<<"Failed to clear cache for "<<name<<std::endl;
```
Comment (mind04) [inline pdns/dbdnsseckeeper.cc:148]: This is dead code. clearKeyCache and clearMetaCache always return true.

Comment (cmouse) [reply pdns/dbdnsseckeeper.cc:148]: wonder why they bother returning value at all? is it some future-planning thing?

Comment (Habbie) [reply pdns/dbdnsseckeeper.cc:148]: as discussed on IRC: the `return addKey(name, dspk, id, active, published) && clearKeyCache(name);` construction that is used in 15 variations throughout dbdnsseckeeper.cc needs the `true` return value.

**Comment by cmouse** [comment]
fixed nits.

File: pdns/dbdnsseckeeper.cc [hunk 5]
```diff
@@ -126,7 +126,8 @@ void DNSSECKeeper::clearAllCaches() {
   s_metacache.clear();
 }
 
-
+/* This function never fails, the return value is to simplify call chains
+   elsewhere to we can do mutate<cache> && clear<cache> */
```
Comment (Habbie) [inline pdns/dbdnsseckeeper.cc:130]:
```suggestion
   elsewhere so we can do mutate<cache> && clear<cache> */
```

File: pdns/dnsdist.cc [hunk 6]
```diff
@@ -1850,9 +1850,12 @@ static void setUpLocalBind(std::unique_ptr<ClientState>& cs)
 
   if(!cs->tcp && IsAnyAddress(cs->local)) {
     int one=1;
-    setsockopt(fd, IPPROTO_IP, GEN_IP_PKTINFO, &one, sizeof(one));     // linux supports this, so why not - might fail on other systems
+
+    if (setsockopt(fd, IPPROTO_IP, GEN_IP_PKTINFO, &one, sizeof(one)) < 0)    // linux supports this, so why not - might fail on other systems
```
Comment (Habbie) [inline pdns/dnsdist.cc:1854]: the comment suggests that this does not warrant a warning

File: pdns/test-svc_records_cc.cc [hunk 7]
```diff
@@ -73,15 +73,8 @@ BOOST_AUTO_TEST_CASE(test_SvcParam_ctor_no_value) {
 BOOST_AUTO_TEST_CASE(test_SvcParam_ctor_value) {
     string val = "foobar";
     auto base64val = Base64Encode(val);
-    SvcParam param;
-
-    BOOST_CHECK_THROW(SvcParam(SvcParam::mandatory, val), std::invalid_argument);
-    BOOST_CHECK_THROW(SvcParam(SvcParam::alpn, val), std::invalid_argument);
-    BOOST_CHECK_THROW(SvcParam(SvcParam::no_default_alpn, val), std::invalid_argument);
```
Comment (Habbie) [inline pdns/test-svc_records_cc.cc:80]: why is this block gone?

File: pdns/logger.cc [hunk 8]
```diff
@@ -120,6 +120,9 @@ void Logger::log(const string &msg, Urgency u) noexcept
       catch (const runtime_error& e) {
         cerr << e.what() << endl;
       }
+      catch (boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<std::length_error> >& e2) {
```
Comment (Habbie) [inline pdns/logger.cc:123]: is there no less specific form we can catch? what throws this?

File: pdns/receiver.cc [hunk 9]
```diff
@@ -566,8 +566,13 @@ int main(int argc, char **argv)
       
       if(::arg().mustDo("control-console"))
         dl=make_unique<DynListener>();
-      else
-        dl=std::unique_ptr<DynListener>(new DynListener(s_programname));
+      else {
+        try {
+           dl=std::unique_ptr<DynListener>(new DynListener(s_programname));
+        } catch (boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::io::too_few_args> >& e) {
```
Comment (Habbie) [inline pdns/receiver.cc:572]: any less specific exception we can catch?

Comment (Habbie) [reply pdns/receiver.cc:572]: where is this exception coming from, anyway? :)

Comment (cmouse) [reply pdns/receiver.cc:572]: it's what coverity complained, so i copied it from there.

Comment (Habbie) [reply pdns/receiver.cc:572]: let's remove it

File: pdns/logger.cc (hunk 8 above)
Comment (cmouse) [reply pdns/logger.cc:123]: ` S.ringAccount("logmessages",msg);`

Comment (cmouse) [reply pdns/logger.cc:123]:
```
1. fun_call_w_exception: Called function throws an exception of type boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<std::length_error> >.
 d_rings[name].account(item);
```

File: pdns/receiver.cc (hunk 9 above)
Comment (cmouse) [reply pdns/receiver.cc:572]: lets remove it indeed, the source for this is impossible

```
throw runtime_error((boost::format("Error: Unable to parse group ID %s") % cstr).str() );
```

File: pdns/dnsdist.cc (hunk 6 above)
Comment (rgacogne) [reply pdns/dnsdist.cc:1854]: Indeed, I don't think we want a warning for this.

**Comment by pieterlexis** [comment]
@cmouse ping

File: pdns/test-svc_records_cc.cc (hunk 7 above)
Comment (cmouse) [reply pdns/test-svc_records_cc.cc:80]: too aggressive cleanup =)

File: pdns/logger.cc (hunk 8 above)
Comment (cmouse) [reply pdns/logger.cc:123]: ping @Habbie

Comment (Habbie) [reply pdns/logger.cc:123]: If you can put a comment in there explaining why we catch this very specific thing, it's fine by me.

**Review by Habbie — CHANGES_REQUESTED** [review]
One last change requested. After that I'd like a review from at least Otto or Rémi just to be sure.

File: pdns/comfun.cc [hunk 10]
```diff
@@ -69,8 +69,9 @@ struct SendReceive
   {
     d_socket = socket(AF_INET, SOCK_DGRAM, 0);
```
Comment (omoerbeek) [inline pdns/comfun.cc:70]: While here check return `socket()` return val

File: pdns/logger.cc (hunk 8 above)
Comment (cmouse) [reply pdns/logger.cc:123]: I think it is possible to happen if d_rings[name] does not exist, but I wonder if boost would let me catch std::length_error instead.

File: pdns/comfun.cc [hunk 11]
```diff
@@ -69,8 +69,9 @@ struct SendReceive
...
     int val=1;
-    setsockopt(d_socket, SOL_SOCKET, SO_REUSEADDR, &val, sizeof(val));
-    
+    if (setsockopt(d_socket, SOL_SOCKET, SO_REUSEADDR, &val, sizeof(val)) != 0)
+       g_log<<Logger::Warning<<"Failed to turn on SO_REUSEADDR: " << stringerror() << endl;
```
Comment (omoerbeek) [inline pdns/comfun.cc:73]: Are yo sure this is not going to spam the logs if it fails. Is it really useful to know?

File: pdns/comfun.cc [hunk 12]
```diff
@@ -198,8 +199,10 @@ struct SendReceiveRes
   {
     d_socket = socket(AF_INET, SOCK_DGRAM, 0);
     int val=1;
-    setsockopt(d_socket, SOL_SOCKET, SO_REUSEADDR, &val, sizeof(val));
-    connect(d_socket, (struct sockaddr*)&remote, remote.getSocklen());
+    if (setsockopt(d_socket, SOL_SOCKET, SO_REUSEADDR, &val, sizeof(val)) != 0)
+       g_log<<Logger::Warning<<"Failed to turn on SO_REUSEADDR: " << stringerror() << endl;
+    if (connect(d_socket, (struct sockaddr*)&remote, remote.getSocklen()) != 0)
```
Comment (omoerbeek) [inline pdns/comfun.cc:204]: A connect failure feels fatal to me

File: pdns/comfun.cc [hunk 13]
```diff
@@ -198,8 +199,10 @@ struct SendReceiveRes
   {
     d_socket = socket(AF_INET, SOCK_DGRAM, 0);
     int val=1;
-    setsockopt(d_socket, SOL_SOCKET, SO_REUSEADDR, &val, sizeof(val));
-    connect(d_socket, (struct sockaddr*)&remote, remote.getSocklen());
+    if (setsockopt(d_socket, SOL_SOCKET, SO_REUSEADDR, &val, sizeof(val)) != 0)
```
Comment (omoerbeek) [inline pdns/comfun.cc:202]: some worry about filling the log here. Also I'm wondering if it is of use to set `SO_REUSEADDR` on a client socket.

File: pdns/distributor.hh [hunk 14]
```diff
@@ -98,6 +98,8 @@ public:
   {
     QuestionData(const Question& query): Q(query)
     {
+       id = 0;
```
Comment (omoerbeek) [inline pdns/distributor.hh:101]: I prefer {} initializers with the fields themselves: nicething is anybody adds a ct and forget the init, things will keep on working.

File: pdns/logger.cc (hunk 8 above)
Comment (omoerbeek) [reply pdns/logger.cc:123]: IMO this is too specific: if boost changes an implementation detail this won't match.

Comment (cmouse) [reply pdns/logger.cc:123]: removed as impossible.

**Review by omoerbeek — APPROVED** [review]

File: pdns/dnsdist.cc [hunk 15]
```diff
@@ -1885,9 +1885,10 @@ static void setUpLocalBind(std::unique_ptr<ClientState>& cs)
 
   if(!cs->tcp && IsAnyAddress(cs->local)) {
     int one=1;
-    setsockopt(fd, IPPROTO_IP, GEN_IP_PKTINFO, &one, sizeof(one));     // linux supports this, so why not - might fail on other systems
+    (void)setsockopt(fd, IPPROTO_IP, GEN_IP_PKTINFO, &one, sizeof(one)); // linux supports this, so why not - might fail on other systems
 #ifdef IPV6_RECVPKTINFO
-    setsockopt(fd, IPPROTO_IPV6, IPV6_RECVPKTINFO, &one, sizeof(one));
+    if (setsockopt(fd, IPPROTO_IPV6, IPV6_RECVPKTINFO, &one, sizeof(one)) < 0)
+        warnlog("Warning: IPV6_RECVPKTINFO setsockopt failed: %s", stringerror());
```
Comment (rgacogne) [inline pdns/dnsdist.cc:1891]: I don't really see the point of warning about this. There is nothing the user can do about it anyway.

**Review by rgacogne — CHANGES_REQUESTED** [review]

File: pdns/dnsdist.cc (hunk 15 above)
Comment (cmouse) [reply pdns/dnsdist.cc:1891]:
```
18:52 < cmouse> rgacogne: what happens if setting IPV6_RECVPKTINFO fails?
18:53 < cmouse> rgacogne: will it break anything? does something stop working?
19:00 < rgacogne> cmouse: we might not be able to properly detect the destination address for datagrams received over IPv6 for sockets bound to ::
19:01 < cmouse> rgacogne: point being; if setting that fails, will it cause something that might concern an admin?
19:01 < Habbie> yes
19:01 < Habbie> if they have :: in local-address
19:01 < cmouse> then I would suggest that warning would be logged
19:01 < Habbie> but
19:02 < Habbie> we really only want that warning once
19:02 < rgacogne> Habbie: it's for dnsdist in that case
19:02 < Habbie> oh
19:02 < rgacogne> anyway there is nothing an admin can do
19:03 < Habbie> except stop listening on ::
19:03 < Habbie> like the pdns faq used to warn
19:03 < rgacogne> and we would log that every time we start
19:03 < rgacogne> true
19:03 < Habbie> but, when would this happen? on some obsecure solaris fork?
19:03 < cmouse> if kernel does not support that option, most likely.
```

so what do we do?

Comment (Habbie) [reply pdns/dnsdist.cc:1891]: Warn once? (maybe have a (`static`?) `bool` to make sure it really is once.)

Comment (rgacogne) [reply pdns/dnsdist.cc:1891]: Warning only once would be nice but please keep it simple, I don't care that much :-)

Comment (cmouse) [reply pdns/dnsdist.cc:1894]: Fixed.

File: pdns/dnsdist.cc [hunk 16]
```diff
@@ -1852,6 +1852,8 @@ static void checkFileDescriptorsLimits(size_t udpBindsCount, size_t tcpBindsCoun
   }
 }
 
+static bool g_warned_ipv6_recvpktinfo = FALSE;
```
Comment (rgacogne) [inline pdns/dnsdist.cc:1855]:
```suggestion
static bool g_warned_ipv6_recvpktinfo = false;
```

File: pdns/dnsdist.cc [hunk 17]
```diff
@@ -1885,9 +1887,13 @@ static void setUpLocalBind(std::unique_ptr<ClientState>& cs)
...
   if(!cs->tcp && IsAnyAddress(cs->local)) {
     int one=1;
-    setsockopt(fd, IPPROTO_IP, GEN_IP_PKTINFO, &one, sizeof(one));     // linux supports this, so why not - might fail on other systems
+    (void)setsockopt(fd, IPPROTO_IP, GEN_IP_PKTINFO, &one, sizeof(one)); // linux supports this, so why not - might fail on other systems
 #ifdef IPV6_RECVPKTINFO
-    setsockopt(fd, IPPROTO_IPV6, IPV6_RECVPKTINFO, &one, sizeof(one));
+    if (unlikely(setsockopt(fd, IPPROTO_IPV6, IPV6_RECVPKTINFO, &one, sizeof(one))) < 0 &&
+        !g_warned_ipv6_recvpktinfo) {
+        warnlog("Warning: IPV6_RECVPKTINFO setsockopt failed: %s", stringerror());
+        g_warned_ipv6_recvpktinfo == TRUE;
```
Comment (rgacogne) [inline pdns/dnsdist.cc:1895]:
```suggestion
        g_warned_ipv6_recvpktinfo = true;
```

File: pdns/dnsdist.cc [hunk 18]
```diff
@@ -1885,9 +1887,13 @@ static void setUpLocalBind(std::unique_ptr<ClientState>& cs)
 
   if(!cs->tcp && IsAnyAddress(cs->local)) {
     int one=1;
-    setsockopt(fd, IPPROTO_IP, GEN_IP_PKTINFO, &one, sizeof(one));     // linux supports this, so why not - might fail on other systems
+    (void)setsockopt(fd, IPPROTO_IP, GEN_IP_PKTINFO, &one, sizeof(one)); // linux supports this, so why not - might fail on other systems
 #ifdef IPV6_RECVPKTINFO
-    setsockopt(fd, IPPROTO_IPV6, IPV6_RECVPKTINFO, &one, sizeof(one));
+    if (unlikely(setsockopt(fd, IPPROTO_IPV6, IPV6_RECVPKTINFO, &one, sizeof(one))) < 0 &&
```
Comment (rgacogne) [inline pdns/dnsdist.cc:1892]: I don't think we have the `unlikely` macro defined?

**Review by rgacogne — CHANGES_REQUESTED** [review]

File: pdns/dnsdist.cc (hunk 18 above)
Comment (cmouse) [reply pdns/dnsdist.cc:1892]: ah, my bad.

File: pdns/comfun.cc (hunk 12 above)
Comment (Habbie) [reply pdns/comfun.cc:208]: This does not look resolved.

**Review by Habbie — CHANGES_REQUESTED** [review]
Found one older comment that was not resolved. Otherwise, this looks fine to me. Please do a rebase on fresh master so this runs in our current CI settings.

File: pdns/comfun.cc (hunk 12 above)
Comment (cmouse) [reply pdns/comfun.cc:208]: so - what should I do here? raise exception?

**Review by Habbie — APPROVED** [review]

**Pull request merged**